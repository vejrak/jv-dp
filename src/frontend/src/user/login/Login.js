// @flow

import React, { Component } from 'react'
import { Button } from 'react-bootstrap'
import { login } from '../actions'
import type { ErrorResponse } from '../../types'
import Spinner from '../../components/Spinner'
import * as validation from '../../helpers/validation'
import Input from '../../components/Input'
import { isFormValid } from '../../helpers'

type Props = $ReadOnly<{|
  login: typeof login,
  isFetching: boolean,
  error: ?ErrorResponse,
|}>

type State = {
  formData: {
    username: string,
    password: string,
  },
  validationResults: validation.ValidationResults<*>,
}

class Login extends Component<Props, State> {
  state = {
    formData: {
      username: '',
      password: '',
    },
    validationResults: {},
  }

  validate = () => {
    const { formData } = this.state
    return {
      username: validation.required(formData.username),
      password: validation.required(formData.password),
    }
  }

  handleChange = (value: Object) => {
    const { formData, validationResults } = this.state
    delete validationResults[Object.keys(value)[0]]
    this.setState({
      formData: { ...formData, ...value },
      validationResults,
    })
  }

  handleClick = () => {
    const { formData } = this.state
    const validationResults = this.validate()
    this.setState({ validationResults })
    if (isFormValid(validationResults)) {
      this.props.login({ ...formData })
    }
  }

  render() {
    const { formData, validationResults } = this.state
    const { isFetching, error } = this.props

    if (isFetching) return <Spinner />

    return (
      <div className="form-group">
        {error && (
          <div className="alert alert-danger" role="alert">
            Wrong username or password
          </div>
        )}
        <label htmlFor="username">Username</label>
        <Input
          className="form-control"
          id="username"
          name="username"
          onChange={({ target }) =>
            this.handleChange({ username: target.value })
          }
          validationError={validationResults.username}
          value={formData.username}
        />
        <label className="mt-2" htmlFor="password">
          Password
        </label>
        <Input
          className="form-control"
          name="password"
          onChange={({ target }) =>
            this.handleChange({ password: target.value })
          }
          type="password"
          validationError={validationResults.password}
          value={formData.password}
        />
        <Button className="mt-4" onClick={this.handleClick}>
          Login
        </Button>
      </div>
    )
  }
}

export default Login
