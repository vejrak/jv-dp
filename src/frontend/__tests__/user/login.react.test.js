import React from 'react'
import renderer from 'react-test-renderer'
import Login from '../../src/user/login/Login'

it('renders correctly', () => {
  const tree = renderer.create(<Login />).toJSON()
  expect(tree).toMatchSnapshot()
})
