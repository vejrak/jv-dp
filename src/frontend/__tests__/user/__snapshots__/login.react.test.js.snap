// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="username"
  >
    Username
  </label>
  <div>
    <input
      className="form-control"
      id="username"
      name="username"
      onChange={[Function]}
      value=""
    />
  </div>
  <label
    className="mt-2"
    htmlFor="password"
  >
    Password
  </label>
  <div>
    <input
      className="form-control"
      name="password"
      onChange={[Function]}
      type="password"
      value=""
    />
  </div>
  <button
    className="mt-4 btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Login
  </button>
</div>
`;
